File: challenges/999/110.BalancedBinaryTree.go
package challenges

/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }

Given a binary tree, determine if it is height-balanced.

For this problem, a height-balanced binary tree is defined as:

a binary tree in which the left and right subtrees of every node differ in height by no more than 1.

Example 1:

Input: root = [3,9,20,null,null,15,7]
Output: true

Example 2:

Input: root = [1,2,2,3,3,null,null,4,4]
Output: false

Example 3:

Input: root = []
Output: true
*/
func isBalanced(root *TreeNode) bool {
	return height(root) >= 0
}

func height(root *TreeNode) int {
	if root == nil {
		return 0
	}

	if root.Left == nil && root.Right == nil {
		return 1
	}

	lh, rh := height(root.Left), height(root.Right)
	if lh < 0 || rh < 0 {
		return -1
	}

	if absDiff(lh, rh) > 1 {
		return -1
	}

	return max(lh, rh) + 1
}

File: challenges/499/110-balanced-binary-tree.py
'''
110 Balanced Binary Tree
'''

from typing import Optional


# Definition for a binary tree node.
class TreeNode:
  def __init__(self, val=0, left=None, right=None):
    self.val = val
    self.left = left
    self.right = right

  
class Solution:
  def isBalanced(self, root: Optional[TreeNode]) -> bool:
    def is_balanced(node):
      if not node:
        return True, 0

      lb, ld = is_balanced(node.left)
      rb, rd = is_balanced(node.right)
      # print('iter:', node.val, ld, rd)

      if not lb or not rb:
        return False, -1

      if abs(ld-rd) > 1:
        return False, -1

      return True, 1+max(ld, rd)

    res, _ = is_balanced(root)

    return res
        